pkg/network/transport/transport.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// Config contains all configuration parameters for a Transport
+type Config struct {
+	PublicKey     ed25519.PublicKey  // Node's public key
+	PrivateKey    ed25519.PrivateKey // Node's private key
+	TLSCert       *tls.Certificate   // TLS certificate
+	ListenAddr    string             // Address to listen on
+	CertValidator CertValidator      // Certificate validator
+	Handler       ConnectionHandler  // Connection handler
+}
+
+// Transport manages QUIC connections and their lifecycles
+type Transport struct {
+	config       Config
+	listener     *quic.Listener
+	mu           sync.RWMutex
+	conns        map[string]*Conn // Active connections mapped by peer key
+	ctx          context.Context
+	cancel       context.CancelFunc
+	acceptLoopWg sync.WaitGroup // For clean shutdown of accept loop
+}
+
+// NewTransport creates and configures a new transport instance.
+// Returns an error if any required configuration is missing or invalid.
+func NewTransport(config Config) (*Transport, error) {
+	if config.TLSCert == nil {
+		return nil, fmt.Errorf("TLS certificate required")
+	}
+	if config.CertValidator == nil {
+		return nil, fmt.Errorf("certificate validator required")
+	}
+	if config.Handler == nil {
+		return nil, fmt.Errorf("connection handler required")
+	}
+
+	// Verify the certificate
+	if err := config.CertValidator.ValidateCertificate(config.TLSCert.Leaf); err != nil {
+		return nil, ErrInvalidCertificate
+	}
+
+	return &Transport{
+		config: config,
+		conns:  make(map[string]*Conn),
+	}, nil
+}
+
+// Start initializes the transport listener and begins accepting connections.
+// Returns an error if starting the listener fails.
+func (t *Transport) Start() error {
+	tlsConfig := &tls.Config{
+		Certificates:       []tls.Certificate{*t.config.TLSCert},
+		NextProtos:         t.config.Handler.GetProtocols(),
+		ClientAuth:         tls.RequireAnyClientCert,
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: true,
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("%w: %v", ErrInvalidCertificate, err)
+			}
+			return t.config.CertValidator.ValidateCertificate(c)
+		},
+	}
+
+	listener, err := quic.ListenAddr(t.config.ListenAddr, tlsConfig, &quic.Config{
+		EnableDatagrams: true,
+		MaxIdleTimeout:  MaxIdleTimeout,
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrListenerFailed, err)
+	}
+
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	t.listener = listener
+	t.acceptLoopWg.Add(1)
+	go func() {
+		defer t.acceptLoopWg.Done()
+		t.acceptLoop()
+	}()
+	return nil
+}
+
+// Stop gracefully shuts down the transport and all active connections.
+// Waits for the accept loop to finish before returning.
+func (t *Transport) Stop() error {
+	t.cancel()
+	if t.listener != nil {
+		t.listener.Close()
+	}
+	t.acceptLoopWg.Wait() // Wait for acceptLoop to finish
```

Comment:
How are we ensuring all the connections get closed aswell? (might have missed this)